p2plending/WeiLend.sol

```diff
@@ -0,0 +1,187 @@
+            l.beneficiary = _beneficiary;
+            // again assign to the loan l the fundingGoal defined by the argument of the function
+            l.fundingGoal = _goal;
+            // again assign to the loan l the timelimit defined by the argument of the function
+            l.timelimit = block.timestamp + (_timelimit_m * 43200);
+            // again assign to the loan l the category defined by the argument of the function
+            l.category = _category;
+            // again assign to the loan l the interest rate defined by the argument of the function
+            l.interestRateM = _interest_rate / 12;
+            // again assign to the loan l the grace period defined by the argument of the function
+            l.gracePeriod = _grace_period_m * 43200;
+            // again assign to the loan l the tenor defined by the argument of the function (note, tenor meant to be after grace period)
+            l.tenorM = _tenor_a * 12;
+            // now creates a user called u, which is the sender of the transaction
+            User u = users[msg.sender];
+            // creates a variable called u_lid, which adds a one to the number of loans the user has made.
+            uint u_lid = u.numLoans++;
+            u.loans[u_lid] = lid;
+            
+            // calls the event onNewLoan
+            onNewLoan(msg.sender, lid);
+        }
+    }
+ 
+ function contribute(uint _lid)
+    {
+        Loan l = loans[_lid]; // Cannot be expired.
+        
+        // if the raising is not yet terminated
+        if(l.timelimit >= block.timestamp) {
+            uint fid = l.numFunders++;
+            Funder f = l.funders[fid];
+            f.addr = msg.sender;
+            f.amount = msg.value;
+            //increases the loan amount by the contribution
+            l.amount += f.amount;
+            //sends a funder id to the funder
+            l.toFunder[msg.sender] = fid;
+            onContribute(msg.sender, _lid, l.amount);
+        }
+    }
+
+ function refund(uint _lid)
+    {
+        Loan l = loans[_lid];
+        if (block.timestamp > l.timelimit 
+        && l.amount < l.fundingGoal && l.amount > 0){
+            Funder f = l.funders[l.toFunder[msg.sender]];
+            if(f.amount > 0){
+                f.addr.send(f.amount);
+                l.amount -= f.amount;
+                onRefund(msg.sender, _lid, f.amount);
+                f.amount = 0;
+            }
+        }
+    }
+
+ function payout(uint _lid)
+    {
+        Loan l = loans[_lid];
+        if (l.amount >= l.fundingGoal){
+            
+            l.beneficiary.send(l.amount);
+            onPayout(msg.sender, _lid, l.amount);
+            
+            l.status = 1;
+            l.timelimit = l.timelimit + l.gracePeriod;
+            //calculates the monthly installment amount. Constant installment loan
+            l.installment = l.amount * ((l.interestRateM*((1+l.interestRateM)^(l.tenorM)))/(((1+l.interestRateM)^(l.tenorM))-1));
```

Comment:
You probably want to use `**` (exponentiation) and not `^` (exclusive bitwise or).